text/0047-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,313 @@
+
+In other words, we want a uniformly distributed, deterministic mapping from `ValidatorIndex` to `ChunkIndex` per block
+per core.
+
+It's desirable to not embed this function in the runtime, for performance and complexity reasons.
+However, this means that the function needs to be kept very simple and with minimal or no external dependencies.
+Any change to this function could result in parachains being stalled and needs to be coordinated via a runtime upgrade
+or governance call.
+
+#### Proposed function
+
+Pseudocode:
+
+```rust
+pub fn get_chunk_index(
+  n_validators: u32,
+  validator_index: ValidatorIndex,
+  block_number: BlockNumber,
+  core_index: CoreIndex
+) -> ChunkIndex {
+  let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
+  let core_start_pos = abs(core_index - block_number) * threshold;
+
+  (core_start_pos + validator_index) % n_validators
+}
+```
+
+### Network protocol
+
+The request-response `/polkadot/req_chunk` protocol will be bumped to a new version (from v1 to v2).
+For v1, the request and response payloads are:
+```rust
+/// Request an availability chunk.
+pub struct ChunkFetchingRequest {
+	/// Hash of candidate we want a chunk for.
+	pub candidate_hash: CandidateHash,
+	/// The index of the chunk to fetch.
+	pub index: ValidatorIndex,
+}
+
+/// Receive a requested erasure chunk.
+pub enum ChunkFetchingResponse {
+	/// The requested chunk data.
+	Chunk(ChunkResponse),
+	/// Node was not in possession of the requested chunk.
+	NoSuchChunk,
+}
+
+/// This omits the chunk's index because it is already known by
+/// the requester and by not transmitting it, we ensure the requester is going to use his index
+/// value for validating the response, thus making sure he got what he requested.
+pub struct ChunkResponse {
+	/// The erasure-encoded chunk of data belonging to the candidate block.
+	pub chunk: Vec<u8>,
+	/// Proof for this chunk's branch in the Merkle tree.
+	pub proof: Proof,
+}
+```
+
+Version 2 will add an `index` field to `ChunkResponse`:
+
+```rust
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct ChunkResponse {
+	/// The erasure-encoded chunk of data belonging to the candidate block.
+	pub chunk: Vec<u8>,
+	/// Proof for this chunk's branch in the Merkle tree.
+	pub proof: Proof,
+	/// Chunk index.
+	pub index: ChunkIndex
+}
+```
+
+An important thing to note is that in version 1, the `ValidatorIndex` value is always equal to the `ChunkIndex`.
+Until the feature is enabled, this will also be true for version 2. However, after the feature is enabled,
+this will generally not be true.
+
+The requester will send the request to validator with index `V`. The responder will map the `V` validator index to the
+`C` chunk index and respond with the `C`-th chunk.
+
+The protocol implementation MAY check the returned `ChunkIndex` against the expected mapping to ensure that
+it received the right chunk.
```

Comment:
In any case the requester has to verify the chunk's correctness using the provided index. Then given that it can not know (if the mapping is not available) whether the received chunk corresponds to the requested index, it has to keep track of received chunk indices and check for duplicates. (Duplicates should be considered the same as an invalid/garbage response (we drop it and move on to the next validator - we can not punish as in reputation changes, because we don't know which validator misbehaved.)

Ideally see above we can make this all more robust with an updated candidate receipt.